hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/DirectoryStateHelper.java

```diff
@@ -0,0 +1,110 @@
+ * With Blob Endpoint support, driver need to handle implicit paths in store at client side.
+ * On DFS Endpoint, this handling is done on server side and driver works seamlessly.
+ * This toll will be used by tests classes to assert that HDFS APIs work
+ * seamlessly on implicit paths even with Blob Endpoint.
+ */
+public class DirectoryStateHelper {
+
+  /**
+   * DFS Endpoint abstracts nature of directory from user and hence there is no
+   * way to detect implicit directory using DFS Endpoint APIs.
+   * To assert that a path exists as implicit directory we need two things to assert.
+   * 1. Blob Endpoint Listing on the path should return some entries.
+   * 2. GetBlobProperties on path should fail on Blob Endpoint.
+   * @param path to be checked. Can be relative or absolute.
+   * @param fs AzureBlobFileSystem for API calls
+   * @return boolean whether the path exists as Implicit directory or not
+   */
+  public static boolean isImplicitDirectory(Path path, AzureBlobFileSystem fs,
+      TracingContext testTracingContext) throws Exception {
+    Assume.assumeTrue(fs.getAbfsStore().getAbfsConfiguration().getFsConfiguredServiceType() == AbfsServiceType.BLOB);
+    path = new Path(fs.makeQualified(path).toUri().getPath());
+    String relativePath = fs.getAbfsStore().getRelativePath(path);
+
+    // Implicit nature can be checked only on Blob Endpoint.
+    AbfsBlobClient client = fs.getAbfsStore().getClientHandler().getBlobClient();
+    AbfsRestOperation op = null;
+    boolean isNotFound, isemptyList = true;
+
+    // 1st condition: listPaths should return some entries.
+    op = client.listPath(relativePath, false, 1, null, testTracingContext);
+    if (op != null && op.getResult() != null) {
+      int listSize = op.getResult().getListResultSchema().paths().size();
+      if (listSize == 0) {
+        isemptyList = true;
+      } else if (listSize > 0) {
+        isemptyList = false;
+      }
+    }
+
+    // 2nd Condition: getPathStatus fails with 404.
+    try {
+      op = client.getPathStatus(relativePath, testTracingContext, null, false);
+      isNotFound = false;
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() == HTTP_NOT_FOUND) {
+        isNotFound = true;
+      } else {
+        isNotFound = false;
+      }
+    }
+
+    return !isemptyList && isNotFound;
+  }
+
+  /**
+   * To assert that a path exists as explicit directory, we need to assert that
+   * marker blob exists on the path for both DFS and Blob Endpoint.
+   * @param path to be checked
+   * @param fs AzureBlobFileSystem for API calls
+   * @return boolean whether the path exists as Implicit directory or not
+   */
+  public static boolean isExplicitDirectory(Path path, AzureBlobFileSystem fs,
+      TracingContext testTracingContext) throws Exception {
+    Assume.assumeTrue(fs.getAbfsStore().getAbfsConfiguration().getFsConfiguredServiceType() == AbfsServiceType.BLOB);
```

Comment:
Why do we need this check here, this function should work for both DFS and BLOB

Reply:
Changed this check to work with both DFS and Blob for FNS only.
HNS does not have this concept and hence skipping for it.